### PR TITLE
persist: use semaphore-based physical flow control in cc replica fetches

### PR DESCRIPTION
**Pull request by danhhz**
Unblock the transition to cc replicas by reintroducing physical
persist_source flow control. We introduce a new mechanism to improve a
couple deficiencies in the previous one:

- (1) Our product principle of any dataflow that works in the steady
  state should be able to rehydrate (see below).
- (2) More directly model the constrained resource.

Thought experiment for (1): A single dataflow has N inputs, each a
unique shard. One by one, a 128 MiB part is written to each of them. If
we want this to rehydrate, the unavoidable conclusion is that we need
cross-persist_source flow control.

- This has historically been stressful because the old mechanism worked
  on granular timely timestamps, which could easily lead to deadlocks if
  used across multiple inputs (thus each was totally isolated).
- We work around this here by passing a release-on-drop permit token to
  the decode operator, which is run-to-completion.

For (2), we directly model the per-process constrained resource
(mem/disk) with a per-process semaphore having `mem_limit_bytes`
permits.

- Persist fetch data is put in lgalloc, so disk is the real limiting
  resource.
- This targets 50% disk utilization (with the standard cc 1:2 ratio)
  from persist, which seems like a reasonable place to start.
- This is also enough that stalls of the fetch pipeline seem unlikely in
  practice, so there's probably not much benefit to targeting more than
  100% of memory.
- Initial experiments have shown impressive results:
  - Flat usage for constrained scenarios (25cc rehydrating tpch indexes)
  - No slowdown for unconstrained scenarios (50cc rehydrating tpch
    indexes)
  - Notably, the previous mechanism achieved _neither_ of these.
  - This may allow us to default the new mechanism to on!

Notes:
- This also unblocks logical persist_source flow control once we've
  moved off cc replicas with the old physical flow control on.
- NB: Having multiple mechanisms in mz for flow control is unfortunate,
  but it seems possible to model this logic using the frontiers
  technique (which Gus's new flow control is also), so it should be
  possible to unify them later. The translation seems tricky though, and
  unblocking cc replicas now is easily worth it if we rip this out.
- This could also easily be made data parallel (adjusting permits
  accordingly). Initially leaning per-process because per-worker could
  result in less efficient resource usage and the cross-worker
  coordination is a rounding error compared to the network requests it
  would accompany.
- I initially tried using one permit per blob, but this led to some
  slowdown of what should have been unconstrained scenarios.
- Uses other than persist_source have the potential for deadlock with
  this new mechanism, so are initially left out. Compaction is already
  bounded memory, so possible we could model that one as acquiring all
  permits upfront? Even if so, leaving this until/when it becomes an
  issue in practice.

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by danhhz** [comment]
Okay @bkirwi @guswynn this should be ready to go!

File: src/persist-client/src/internal/metrics.rs [hunk 1]
```diff
@@ -2417,6 +2422,177 @@ impl BlobMemCache {
...
+        }
+        let cfg = self.cfg.clone();
+        let registry = self.registry.clone();
+        let init = async move {
+            let total_permits = match cfg.announce_memory_limit {
+                // Non-cc replicas have the old physical flow control mechanism,
+                // so only apply this one on cc replicas.
+                Some(mem) if cfg.is_cc_active => {
+                    // We can't easily adjust the number of permits later, so
+                    // make sure we've synced dyncfg values at least once.
```
Comment (bkirwi) [inline src/persist-client/src/internal/metrics.rs:2457]: Depends on your definition of "easy", but it's not a wild amount of code: https://github.com/MaterializeInc/materialize/blob/0fd7267e00ef9bbfb1875a52b88fa16e61311908/src/adapter/src/webhook.rs#L392-L419

File: src/persist-client/src/internal/metrics.rs [hunk 2]
```diff
@@ -2417,6 +2422,177 @@ impl BlobMemCache {
...
+                    total_permits
+                }
+                Some(_) | None => Semaphore::MAX_PERMITS,
+            };
+            MetricsSemaphore::new(&registry, "fetch", total_permits)
+        };
+        self.fetch.get_or_init(|| init).await
+    }
+
+    pub(crate) async fn acquire_fetch_permits(&self, encoded_size_bytes: usize) -> MetricsPermits {
```
Comment (bkirwi) [inline src/persist-client/src/internal/metrics.rs:2473]: Will this get stuck if we request more permits than the total number of permits? (If so, seems safish to cap `requested_permits` at `total_permits`.)

**Review by bkirwi — COMMENTED** [review]
Very pleasantly lightweight!

File: src/persist-client/src/internal/metrics.rs (hunk 1 above)
Comment (danhhz) [reply src/persist-client/src/internal/metrics.rs:2457]: Didn't know we had an example in the repo, but I did think of that technique. The issue is that tokio semaphore has a fixed upper bound on the number of permits (I _think_), which means if you set the dyncfg from big value (e.g. "off") to small value, and then back to big value, it may not be possible to add enough permits for the second big value.

File: src/persist-client/src/internal/metrics.rs (hunk 2 above)
Comment (danhhz) [reply src/persist-client/src/internal/metrics.rs:2473]: It does! Just happens in `acquire_permits`

**Comment by danhhz** [comment]
@bkirwi Friendly ping! Would love to get this merged today

File: src/persist-client/src/internal/metrics.rs (hunk 1 above)
Comment (bkirwi) [reply src/persist-client/src/internal/metrics.rs:2457]: There is definitely...

```rust
 pub(crate) const MAX_PERMITS: usize = std::usize::MAX >> 3;
```

But my understanding is that anything that's an illegal value for `add_permits` is also an illegal value for the initial number of permits, so I think if your example flow breaks it would have broken on the initial big value as well?

(Not critical for this PR, but I do want to make sure I'm understanding right...)

File: src/persist-client/src/internal/metrics.rs (hunk 2 above)
Comment (bkirwi) [reply src/persist-client/src/internal/metrics.rs:2473]: 🤦!!

File: src/persist-txn/src/operator.rs [hunk 3]
```diff
@@ -566,7 +566,7 @@ impl DataSubscribe {
             let (mut data, mut txns) = (ProbeHandle::new(), ProbeHandle::new());
             let data_stream = data_stream.flat_map(|part| {
                 let part = part.parse();
-                part.map(|((k, v), t, d)| {
```
Comment (bkirwi) [inline src/persist-txn/src/operator.rs:569]: Since we own `part` here, we should be able to grab the inner part and skip the clone.

(Personally I would be tempted to avoid the `Deref` implementations and either add explicit methods or make the fields public. But if you prefer to keep the `Deref`s an `From` impl would also do this job!)

**Review by bkirwi — APPROVED** [review]
It's good!

File: src/persist-client/src/internal/metrics.rs (hunk 1 above)
Comment (danhhz) [reply src/persist-client/src/internal/metrics.rs:2457]: Not sure I was completely clear, so let me try again. I could be wrong, but this is my read of the docs:
- dyncfg: MAX_PERMITS -> starts the semaphore with `Semaphore::MAX_PERMITS`
- dyncfg: SMALL -> `semaphore.forget_permits(MAX_PERMITS - SMALL)`. This doesn't reduce the number of permits AFAIK, it just leaks them
- dyncfg: MAX_PERMITS -> `semaphore.add_permits(MAX_PERMITS-SMALL)`. This doesn't work because it already had the maximum number of permits from the first step

Comment (bkirwi) [reply src/persist-client/src/internal/metrics.rs:2457]: Ah, okay!

The semaphore doesn't explicitly track a total number of permits... just a count of available permits. Acquiring a permit decrements the count. Dropping the permit increments it again. (So to the extent which there is a total number of permits, it's the available count plus all the live permits.)

So while forgetting doesn't decrement the semaphore per se, that's okay because it was already decremented when those permits were acquired. However, forgetting _does_ ensure that those permits are never added back to the pool. So the forget-then-add gets you back to the original state.

Playground demo of this not exploding: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1a3fd33701bfe1a8105e395ee0e790f6

Comment (danhhz) [reply src/persist-client/src/internal/metrics.rs:2457]: Aha! Okay then indeed my read of the docs was wrong, glad we shook this out! (cc @guswynn who I misinformed about this.)

I might do this one as a fast-follow if that's okay. I've done a good amount of testing of the current code, and this is something I'd want to re-deploy to my staging env and play around with before merging

File: src/persist-txn/src/operator.rs (hunk 3 above)
Comment (danhhz) [reply src/persist-txn/src/operator.rs:569]: Done!

**Review by danhhz — COMMENTED** [review]
TFTR! Gonna let this nightlies finish and then merge!
